usr/src/test/zfs-tests/tests/functional/cli_root/zfs_program/zfs_program_json.ksh

```diff
@@ -0,0 +1,132 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy is of the CDDL is also available via the Internet
+# at http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2018 Datto Inc.
+#
+
+. $STF_SUITE/usr/src/common/zfs/libtest.shlib
+
+#
+# DESCRIPTION:
+#
+# STRATEGY:
+#	1. Compare JSON output formatting for a channel program to template
+#	2. Using bad command line option (-Z) gives correct error output
+#
+
+verify_runnable "both"
+
+function cleanup
+{
+	log_must zfs destroy $TESTDS
+	return 0
+}
+log_onexit cleanup
+
+log_assert "Channel programs output valid JSON"
+
+TESTDS="$TESTPOOL/zcp-json"
+log_must zfs create $TESTDS
+
+TESTZCP="/$TESTDS/zfs_rlist.zcp"
+cat > "$TESTZCP" << EOF
+	succeeded = {}
+	failed = {}
+
+	function list_recursive(root, prop)
+		for child in zfs.list.children(root) do
+			list_recursive(child, prop)
+		end
+		val, src  = zfs.get_prop(root, prop)
+		if (val == nil) then
+			failed[root] = val
+		else
+			succeeded[root] = val
+		end
+	end
+
+	args = ...
+
+	argv = args["argv"]
+
+	list_recursive(argv[1], argv[2])
+
+	results = {}
+	results["succeeded"] = succeeded
+	results["failed"] = failed
+	return results
+EOF
+
+# 1. Compare JSON output formatting for a channel program to template
+typeset -a pos_usr/src/cmds=("recordsize" "type")
+typeset -a pos_usr/src/cmds_out=(
+"{
+    \"return\": {
+        \"failed\": {},
+        \"succeeded\": {
+            \"$TESTDS\": 131072
+        }
+    }
+}"
+"{
+    \"return\": {
+        \"failed\": {},
+        \"succeeded\": {
+            \"$TESTDS\": \"filesystem\"
+        }
+    }
+}")
+typeset -i cnt=0
+typeset usr/src/cmd
+for usr/src/cmd in ${pos_usr/src/cmds[@]}; do
+	log_must zfs program $TESTPOOL $TESTZCP $TESTDS $usr/src/cmd 2>&1
+	log_must zfs program $TESTPOOL -j $TESTZCP $TESTDS $usr/src/cmd 2>&1
+	# json.tool is needed to guarantee consistent ordering of fields
+	# sed is needed to trim trailing space in CentOS 6's json.tool output
+	OUTPUT=$(zfs program $TESTPOOL -j $TESTZCP $TESTDS $usr/src/cmd 2>&1 | python -m json.tool | sed 's/[[:space:]]*$//')
```

Comment:
[nit] Not sure if we do style checks for shell scripts too but can you maybe wrap this line?